### PR TITLE
Add a versioning doc to the repo.

### PR DESCRIPTION
**Pull request by surfnerd**
### Proposed change(s)

Add a versioning doc to the repo

### Types of change(s)

- [ ] Bug fix
- [ ] New feature
- [ ] Code refactor
- [ ] Breaking change
- [x] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

**Comment by surfnerd** [comment]
I still need to link to this from the README

File: docs/Versioning.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+tools will not be versioned.
+
+### Unity package
+Package name: com.unity.ml-agents
+- Versioned following [Semantic Versioning Guidelines](https://www.semver.org)
+- This package produces one artifact, the `.demo` files.  These files will have integer
+    versioning. This means their version will increment by 1 at each change.  The
+    com.unity.ml-agents package must be backward compatible with version changes
+    that occur between minor versions.  For example, consider that com.unity.ml-agents
+    is at version 1.0.0 and the demo files are at version one.  If the demo files
```
Comment (mmattar) [inline docs/Versioning.md:20]: Do we ever read .demo files in the C# package or only write? I thought we generated them in C# and consumed them in Python.

File: docs/Versioning.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+Package names: ml-agents / ml-agents-envs / gym-unity
+- The python packages remain in "Beta."  This means that breaking changes to the public
+    API of the python packages can change without having to have a major version bump.
+    Historically, the python and C# packages were in version lockstep.  This is no longer
+    the case.  The python packages will remain in lockstep in terms of version, while the
+    C# package will follow its own versioning as is appropriate.
+- While the python packages will remain in Beta for now, we acknowledge that the most
+    heavily used portion of our python interface is the `mlagents-learn` CLI and strive
+    to make this part of our API backward compatible. We are actively working on this and
+    expect to have a stable CLI in the next few months.
```
Comment (mmattar) [inline docs/Versioning.md:43]: months --> weeks

File: docs/Versioning.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+Packages which communicate: com.unity.ml-agents / ml-agents-envs
+
+Another entity of the ML-Agents Toolkit that requires versioning is the communicator
+version, which will follow semantic versioning.  This guarantees a level of backward
+compatibility between different versions of C# and Python packages which communicate.
+Any Communicator version 1.x.x of the Unity package should be compatible with any 1.x.x
+Communicator Version in Python.
+
+See [Communication Protocol Versioning](https://docs.google.com/document/d/1gKn4BX5CWfsG_iv_fMnrQGL9FGG7mdUIYCPXsVOe-Ec/edit#heading=h.v7t1ynb2u4tr)
```
Comment (mmattar) [inline docs/Versioning.md:55]: This doc won't be publically viewable. Can we summarize here?

File: docs/Versioning.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+Side Channels are not versioned, but have been designed to support backward compatibility
+for what they are. As of today, we provide 4 side channels:
+- FloatProperties: shared float data between Unity - Python (bidirectional)
+- RawBytes: raw data that can be sent Unity - Python (bidirectional)
+- EngineConfig: a set of numeric fields in a pre-defined order sent from Python to Unity
+- Stats: (name, value, agg) messages sent from Unity to Python
+
+Aside from the specific implementations of side channels we provide (and use ourselves),
+the Side Channel interface is made available for users to create their own custom side
+channels. As such, we need to ensure that interface between Unity and Python is consistent
```
Comment (mmattar) [inline docs/Versioning.md:71]: "we need to ensure" - I think we should use more active tone to indicate what we will and will not support/enable.

File: docs/Versioning.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
+# ML-Agents Versioning
```
Comment (mmattar) [inline docs/Versioning.md:1]: Can we add a subsection that describes our GitHub Releases and the numbering scheme.

File: docs/Versioning.md (hunk 1 above)
Comment (chriselion) [reply docs/Versioning.md:20]: The only time we read them is in the Demonstration Drawer, and that's just summary stats.

File: docs/Versioning.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    supported through the 1.0.0 lifecycle.  We intend to provide stability for our
+    users by moving to a 1.0.0 release of com.unity.ml-agents.
+
+
+### Python Packages
+Package names: ml-agents / ml-agents-envs / gym-unity
+- The python packages remain in "Beta."  This means that breaking changes to the public
+    API of the python packages can change without having to have a major version bump.
+    Historically, the python and C# packages were in version lockstep.  This is no longer
+    the case.  The python packages will remain in lockstep in terms of version, while the
```
Comment (chriselion) [inline docs/Versioning.md:38]: "... will remain in lockstep with each other ..."?

File: docs/Versioning.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+- This package produces one artifact, the `.demo` files.  These files will have integer
+    versioning. This means their version will increment by 1 at each change.  The
+    com.unity.ml-agents package must be backward compatible with version changes
+    that occur between minor versions.  For example, consider that com.unity.ml-agents
+    is at version 1.0.0 and the demo files are at version one.  If the demo files
+    change to version 2, the next release of com.unity.ml-agents at version 1.1.0
+    shall be able to read both of these formats.  If the demo files were to change to
+    version 3 and com.unity.ml-agents to version 2.0.0, support for demo versions 1 and
+    2 could be dropped for com.unity.ml-agents version 2.0.0.
+- This package consumes an artifact of the training process: the `.nn` file.  The
```
Comment (chriselion) [inline docs/Versioning.md:25]: These are integer-versioned too; currently on version 2.

File: docs/Versioning.md [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    com.unity.ml-agents package will need to support the version of `.nn` files
+    which existed at its 1.0.0 release.
```
Comment (chriselion) [inline docs/Versioning.md:27]: "... for the remainder of the 1.0.0 release cycle"? This makes it sounds like we'll support the current format indefinitely.

File: docs/Versioning.md (hunk 1 above)
Comment (surfnerd) [reply docs/Versioning.md:20]: you are correct

File: docs/Versioning.md (hunk 3 above)
Comment (surfnerd) [reply docs/Versioning.md:55]: ah, yeah, for some reason i thought this was an MD file.

File: docs/Versioning.md (hunk 8 above)
Comment (surfnerd) [reply docs/Versioning.md]: change the wording a bit, let me know if it's better.

File: docs/Versioning.md (hunk 1 above)
Comment (surfnerd) [reply docs/Versioning.md]: i used the NN files as an example instead.

File: docs/Versioning.md [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    to be supported for 1.x.x versions of com.unity.ml-agents.  We intend to provide stability
+    for our users by moving to a 1.0.0 release of com.unity.ml-agents.
+
+
+### Python Packages
+Package names: ml-agents / ml-agents-envs / gym-unity
+- The python packages remain in "Beta."  This means that breaking changes to the public
+    API of the python packages can change without having to have a major version bump.
+    Historically, the python and C# packages were in version lockstep.  This is no longer
+    the case.  The python packages will remain in lockstep with each other, while the
```
Comment (chriselion) [inline docs/Versioning.md:55]: Sorry, missed this before. The python packages versions are currently in sync, but might go out of sync in the future (specifically, we might make mlagents_envs 1.0 while the trainers stay at 0.x).

Comment (surfnerd) [reply docs/Versioning.md]: called out that they may diverge in the future.

File: docs/Versioning.md [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+we use to version our packages and the data that flows into, through, and out of them clearly.
+Our project now has four packages (1 Unity, 3 Python) along with artifacts that are produced as
+well as consumed.  This document covers the versioning for these packages and artifacts.
+
+## GitHub Releases
+Up until now, all packages were in lockstep in-terms of versioning. As a result, the GitHub releases
+were tagged with the version of all those packages (e.g. v0.15.0, v0.15.1) and labeled accordingly.
+With the decoupling of package versions, we now need to revisit our GitHub release tagging.
+The proposal is that we move towards an integer release numbering for our repo and each such
+release will call out specific version upgrades of each package. For instance, with the May release,
```
Comment (mmattar) [inline docs/Versioning.md:14]: We actually released on April 30. Maybe we should directly link to those release notes.

File: docs/Versioning.md [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+we will have:
+- GitHub Release 1 (branch name: “release_1_branch)
```
Comment (mmattar) [inline docs/Versioning.md:16]: Missing ending " after release_1_branch. Or italicize.

**Pull request merged**